notes/smart-pointers-1.md

```diff
@@ -0,0 +1,244 @@
+
+```
+
+---
+## Bug #1
+
+We've seen that you can't copy unique pointers, but you can still wrap the same raw pointer with multiple unique pointers. This causes a problem
+
+Example:
+```C++
+int main() {
+    int* p = new int(333);
+    std::unique_ptr<int> x(p);
+    std::unique_ptr<int> y(p);
+
+    // What happens here at the end of main?
+}
+```
+- Now both `x` and `y` "own" pointer `p`
+- When `y` falls out of scope, it will call `delete` on `p`, which it owns
+- But then when `x` falls out of scope, it also will call `delete` on `p`, which has already been deleted!
+- Causes a double-free!
+
+---
+## Bug #2
+
+Also remember that smart pointers are intended to wrap raw pointers to HEAP-ALLOCATED memory
+
+You should never try to `delete` memory on the stack anyways
+
+Bugged example:
+```C++
+int main(){
+  int x = 333;
+  std::unique_ptr<int> p(&x); // BUG
+}
+```
+- This will still compile
+- However, `&x` is a pointer to memory on the stack, so trying to `delete` it will cause undefined behavior
+
+
+---
+## Transferring ownership
+
+You can transfer ownership of a pointer from one `unique_ptr` to another using `release()` and `reset()` in combination
+
+```C++
+int main() {
+    std::unique_ptr<int> x(new int(333));
+
+    std::unique_ptr<int> y(x.release());
+    // y now owns x's old pointer
+    x.reset(new int(124));
+
+    std::cout << *x << std::endl; // prints 124
+    std::cout << *y << std::endl; // prints 333
+
+}
+```
+
+You can also use move semantics (will cover later)
+
+```C++
+std::unique_ptr<int> x(new int(333));
+std::unique_ptr<int> y = std::move(x);
+
+std::cout << x.get() << std::endl; // prints 0
+std::cout << y.get() << std::endl; // prints something like 0x602000000010
+```
+
+---
+## `unique_ptr` and STL
+
+Unique pointers can be stored inside STL containers
+- Since `unique_ptr` supports move semantics, STL will move rather than copy
+
+However if you try to sort, say a vector of unique pointers, you'll be sorting based on pointer address comparison, not the pointed-to objects
+
+Instead, provide a custom comparator function
+
+```C++
+bool cmp(const unique_ptr<int> &x, const unique_ptr<int> &y) {
+  return *x < *y;
+}
+
+void printFunction(unique_ptr<int> &x){
+  cout << *x << endl;
+}
+
+int main(){
+  vector<unique_ptr<int>> vec;
+  vec.push_back(unique_ptr<int>(new int(333)));
+  vec.push_back(unique_ptr<int>(new int(124)));
+
+  // Warning: don't do this
+  sort(vec.begin(), vec.end());
+  // Do this
+  sort(vec.begin(), vec.end(), &cmp);
+  for_each(vec.begin(), vec.end(), &printFunction);
+}
+```
+
+---
+## `unique_ptr` and arrays
+
+One last thing about unique pointers- they can also wrap arrays
+
+```C++
+std::unique_ptr<int[]> x(new int[5]);
+```
+
+When `x` is destroyed, it will call `delete[]` appropriately on its owned pointer
+
+Note: instead of using a `std::unique_ptr` for arrays, it might be better to use `std::array` or `std::vector`
+
+---
+## Conversion to `std::shared_ptr`
+
```

Comment:
Perhaps moving conversions to a whole new section might be best, since it bears no importance to the reader at this stage.